sympy/matrices/common.py

```diff
@@ -726,78 +726,99 @@ def diag(kls, *args, **kwargs):
+            if hasattr(m, 'rows'):
+                return m.rows, m.cols
+            if isinstance(m, dict):
+                return m.get('size', (0, 0))
+            return 1, 1
         def minsize(d):
             r = max(0, -min(d)) + 1
             c = -min(0, -max(d)) + 1
             for k in d:
                 size = len(d[k]) if type(d[k]) is list else 0
                 r = max(r, abs(k) + size)
                 c = max(c, abs(k) + size)
-            return r, c
-
-        def size(m):
-            """Compute the size of the diagonal block"""
-            if hasattr(m, 'rows'):
-                return m.rows, m.cols
-            if isinstance(m, dict):
-                rv = m.pop('size', (kwargs.get('rows'), kwargs.get('cols')))
-                if None in rv:
+            if 'size' in d:
+                R, C = d['size']
+                if r > R or c > C:
                     raise ValueError(filldedent('''
-                        If size is not given in dict the `rows`
-                        and `cols` must be given as args to diag.'''))
-                return rv
-            return 1, 1
+                        the dict size is too small
+                        for the given elements'''))
+                r, c = R, C
+            return r, c
 
-        dict_sizes = []
-        diag_rows = diag_cols = 0
+        # collapse contiguous dicts
         newargs = []
-        nosize = 0
-        num_dicts = 0
+        nosize = []
+        i = 0
         for a in args:
             if newargs and all(isinstance(i, dict) and 'size' not in i for i in (a, newargs[-1])):
                 newargs[-1].update(a)
             else:
                 newargs.append(a)
                 if isinstance(a, dict) and 'size' not in a:
-                    nosize += 1
-            if isinstance(a, dict):
-                num_dicts += 1
-        if nosize > 1:
+                    nosize.append(i)
+                i += 1
+        if len(nosize) > 1:
             raise ValueError(filldedent('''
-                non-contiguous dictionaries must have a size specified,
-                e.g. {'size': (rows, cols), ...}'''))
+                Only one non-contiguous dictionary can have
+                unspecified size. For the others, give the size,
+                e.g. {'size': (rows, cols), ...}.'''))
+        elif nosize:  # length of 1
+            i = nosize[0]
+            if None is (rows, cols):
+                # the unspecified dict will have minimal size
+                dsize = minsize(newargs[i])
+            else:
+                # the unspecified dict will take up
+                # the remaining space
+                spec_rows = spec_cols = 0
+                for a in newargs:
+                    r, c = size(a)
+                    spec_rows += r
+                    spec_cols += c
+                rneed = rows - spec_rows
+                cneed = cols - spec_cols
+                r, c = dsize
```

Comment:
dsize is not assigned any value. It show error in following case.

`Matrix.diag({1:2,-2:1},4, rows = 3, cols = 4)`
I've added assignment of it.